File: src/0111/01_workspace.py
from path import Path


class Config:
    def __init__(self, project_list):
        self.project_list = project_list

    def __repr__(self):
        return "<Config>"


class Workspace:
    """ A workspace contains a collection of projects
    and a config

    """

    def __init__(self, root_path):
        self.root_path = root_path
        self.config = Config(project_list=["foo", "bar"])
        self.projects = []
        for project_name in self.config.project_list:
            project = self.build_project(project_name)
            self.projects.append(project)

    def build_project(self, name):
        project_path = self.root_path / name
        return Project(name, path=project_path)

    def __repr__(self):
        project_list = ",".join(repr(x) for x in self.projects)
        return f"<Workspace in {self.root_path} with projects {project_list}>"


class Project:
    def __init__(self, name, *, path):
        self.name = name
        self.path = path

    def __repr__(self):
        return f"<Project {self.name}>"


def main():
    workspace = Workspace(Path("path/to/workspace"))
    print(workspace)


if __name__ == "__main__":
    main()

File: src/0111/01_workspace.rs
use std::fmt::Display;
use std::path::{Path, PathBuf};

fn parse_config(_root_path: &Path) -> Config {
    Config {
        projects: vec!["foo".to_string(), "bar".to_string()],
    }
}

#[derive(Debug)]
struct Config {
    projects: Vec<String>,
}

#[derive(Debug)]
struct Project {
    name: String,
    path: PathBuf,
}

impl Display for Project {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "<Project {}>", self.name)
    }
}

#[derive(Debug)]
struct Workspace {
    root_path: PathBuf,
    config: Config,
    projects: Vec<Project>,
}

impl Workspace {
    fn new(root_path: &Path) -> Workspace {
        let config = parse_config(root_path);
        // this clone() is a hint something is wrong
        let project_names = config.projects.clone();
        let mut res = Workspace {
            root_path: root_path.to_path_buf(),
            projects: vec![],
            config: config,
        };
        res.projects = project_names.iter().map(|x| res.new_project(x)).collect();
        res
    }

    fn new_project(&self, name: &str) -> Project {
        Project {
            name: name.to_string(),
            path: self.root_path.join(name),
        }
    }
}

fn main() {
    let root_path = PathBuf::from("path/to/Workspace");
    let workspace = Workspace::new(&root_path);
    dbg!(&workspace);
}
